File: rust/autonomi/src/pointer.rs
use autonomi::pointer::{
    Pointer as AutonomiPointer, PointerAddress as AutonomiPointerAddress,
    PointerTarget as AutonomiPointerTarget,
};
use std::sync::Arc;

use crate::data::ChunkAddress;
use crate::keys::{PublicKey, SecretKey};

/// Error type for pointer operations
#[derive(Debug, uniffi::Error, thiserror::Error)]
pub enum PointerError {
    #[error("Invalid pointer: {reason}")]
    InvalidPointer { reason: String },
    #[error("Parsing failed: {reason}")]
    ParsingFailed { reason: String },
}

/// Address of a pointer on the network, derived from the owner's public key
#[derive(uniffi::Object, Clone, Copy, Debug)]
pub struct PointerAddress {
    pub(crate) inner: AutonomiPointerAddress,
}

#[uniffi::export]
impl PointerAddress {
    /// Create a new PointerAddress from a public key (owner)
    #[uniffi::constructor]
    pub fn new(public_key: Arc<PublicKey>) -> Arc<Self> {
        Arc::new(Self {
            inner: AutonomiPointerAddress::new(public_key.inner),
        })
    }

    /// Create a PointerAddress from a hex string
    #[uniffi::constructor]
    pub fn from_hex(hex: String) -> Result<Arc<Self>, PointerError> {
        let inner =
            AutonomiPointerAddress::from_hex(&hex).map_err(|e| PointerError::ParsingFailed {
                reason: format!("Failed to parse hex: {}", e),
            })?;
        Ok(Arc::new(Self { inner }))
    }

    /// Returns the owner's public key
    pub fn owner(&self) -> Arc<PublicKey> {
        Arc::new(PublicKey {
            inner: *self.inner.owner(),
        })
    }

    /// Serialize to hex string
    pub fn to_hex(&self) -> String {
        self.inner.to_hex()
    }
}

/// A mutable pointer to data on the network
/// Stored at the owner's public key address and can only be updated by the owner
#[derive(uniffi::Object, Clone, Debug)]
pub struct NetworkPointer {
    pub(crate) inner: AutonomiPointer,
}

#[uniffi::export]
impl NetworkPointer {
    /// Create a new pointer signed with the provided secret key
    /// There can only be one pointer per key on the network
    #[uniffi::constructor]
    pub fn new(key: Arc<SecretKey>, counter: u64, target: Arc<PointerTarget>) -> Arc<Self> {
        Arc::new(Self {
            inner: AutonomiPointer::new(&key.inner, counter, target.inner.clone()),
        })
    }

    /// Get the network address where this pointer is stored
    pub fn address(&self) -> Arc<PointerAddress> {
        Arc::new(PointerAddress {
            inner: self.inner.address(),
        })
    }

    /// Get the target that this pointer points to
    pub fn target(&self) -> Arc<PointerTarget> {
        Arc::new(PointerTarget {
            inner: self.inner.target().clone(),
        })
    }

    /// Get the counter value (version) of this pointer
    pub fn counter(&self) -> u64 {
        self.inner.counter()
    }
}

/// The target that a pointer can point to on the network
#[derive(uniffi::Object, Clone, Debug)]
pub struct PointerTarget {
    pub(crate) inner: AutonomiPointerTarget,
}

#[uniffi::export]
impl PointerTarget {
    /// Create a pointer target pointing to a chunk
    #[uniffi::constructor]
    pub fn chunk(addr: Arc<ChunkAddress>) -> Arc<Self> {
        Arc::new(Self {
            inner: AutonomiPointerTarget::ChunkAddress(addr.inner),
        })
    }

    /// Create a pointer target pointing to another pointer
    #[uniffi::constructor]
    pub fn pointer(addr: Arc<PointerAddress>) -> Arc<Self> {
        Arc::new(Self {
            inner: AutonomiPointerTarget::PointerAddress(addr.inner),
        })
    }

    /// Serialize to hex string
    pub fn to_hex(&self) -> String {
        self.inner.to_hex()
    }
}

File: csharp/AntFfi/Pointer.cs
using AntFfi.Native;

namespace AntFfi;

/// <summary>
/// Exception thrown when a pointer operation fails.
/// </summary>
public class PointerException : AntFfiException
{
    public PointerException(string message) : base(message) { }
    public PointerException(string message, RustCallStatus status) : base(message, status) { }
}

/// <summary>
/// Address of a pointer on the network (derived from owner's public key).
/// </summary>
public sealed class PointerAddress : NativeHandle
{
    internal PointerAddress(IntPtr handle) : base(handle) { }

    /// <summary>
    /// Creates a PointerAddress from a PublicKey.
    /// </summary>
    public static PointerAddress FromPublicKey(PublicKey publicKey)
    {
        ArgumentNullException.ThrowIfNull(publicKey);
        publicKey.ThrowIfDisposed();

        var status = RustCallStatus.Create();
        var handle = NativeMethods.PointerAddressNew(publicKey.CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PointerAddress.FromPublicKey");
        return new PointerAddress(handle);
    }

    /// <summary>
    /// Creates a PointerAddress from a hex string.
    /// </summary>
    public static PointerAddress FromHex(string hex)
    {
        var hexBuffer = UniFFIHelpers.StringToRustBuffer(hex);
        var status = RustCallStatus.Create();
        var handle = NativeMethods.PointerAddressFromHex(hexBuffer, ref status);
        if (status.IsError)
            throw new PointerException("Failed to parse pointer address from hex", status);
        return new PointerAddress(handle);
    }

    /// <summary>
    /// Gets the owner's PublicKey.
    /// </summary>
    public PublicKey Owner()
    {
        ThrowIfDisposed();
        var status = RustCallStatus.Create();
        var handle = NativeMethods.PointerAddressOwner(CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PointerAddress.Owner");
        return new PublicKey(handle);
    }

    /// <summary>
    /// Returns the hex representation of this pointer address.
    /// </summary>
    public string ToHex()
    {
        ThrowIfDisposed();
        var status = RustCallStatus.Create();
        NativeMethods.PointerAddressToHex(out var buffer, CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PointerAddress.ToHex");
        return UniFFIHelpers.StringFromRustBuffer(buffer);
    }

    /// <inheritdoc/>
    protected override void FreeHandle()
    {
        var status = RustCallStatus.Create();
        NativeMethods.FreePointerAddress(Handle, ref status);
    }

    /// <inheritdoc/>
    protected internal override IntPtr CloneHandle()
    {
        var status = RustCallStatus.Create();
        return NativeMethods.ClonePointerAddress(Handle, ref status);
    }
}

/// <summary>
/// Target of a network pointer - can point to chunks, other pointers, graph entries, or scratchpads.
/// </summary>
public sealed class PointerTarget : NativeHandle
{
    internal PointerTarget(IntPtr handle) : base(handle) { }

    /// <summary>
    /// Creates a pointer target pointing to a chunk.
    /// </summary>
    public static PointerTarget ToChunk(ChunkAddress address)
    {
        ArgumentNullException.ThrowIfNull(address);
        address.ThrowIfDisposed();

        var status = RustCallStatus.Create();
        var handle = NativeMethods.PointerTargetChunk(address.CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PointerTarget.ToChunk");
        return new PointerTarget(handle);
    }

    /// <summary>
    /// Creates a pointer target pointing to another pointer.
    /// </summary>
    public static PointerTarget ToPointer(PointerAddress address)
    {
        ArgumentNullException.ThrowIfNull(address);
        address.ThrowIfDisposed();

        var status = RustCallStatus.Create();
        var handle = NativeMethods.PointerTargetPointer(address.CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PointerTarget.ToPointer");
        return new PointerTarget(handle);
    }

    /// <summary>
    /// Creates a pointer target pointing to a graph entry.
    /// </summary>
    public static PointerTarget ToGraphEntry(GraphEntryAddress address)
    {
        ArgumentNullException.ThrowIfNull(address);
        address.ThrowIfDisposed();

        var status = RustCallStatus.Create();
        var handle = NativeMethods.PointerTargetGraphEntry(address.CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PointerTarget.ToGraphEntry");
        return new PointerTarget(handle);
    }

    /// <summary>
    /// Creates a pointer target pointing to a scratchpad.
    /// </summary>
    public static PointerTarget ToScratchpad(ScratchpadAddress address)
    {
        ArgumentNullException.ThrowIfNull(address);
        address.ThrowIfDisposed();

        var status = RustCallStatus.Create();
        var handle = NativeMethods.PointerTargetScratchpad(address.CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PointerTarget.ToScratchpad");
        return new PointerTarget(handle);
    }

    /// <summary>
    /// Returns the hex representation of this pointer target.
    /// </summary>
    public string ToHex()
    {
        ThrowIfDisposed();
        var status = RustCallStatus.Create();
        NativeMethods.PointerTargetToHex(out var buffer, CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PointerTarget.ToHex");
        return UniFFIHelpers.StringFromRustBuffer(buffer);
    }

    /// <inheritdoc/>
    protected override void FreeHandle()
    {
        var status = RustCallStatus.Create();
        NativeMethods.FreePointerTarget(Handle, ref status);
    }

    /// <inheritdoc/>
    protected internal override IntPtr CloneHandle()
    {
        var status = RustCallStatus.Create();
        return NativeMethods.ClonePointerTarget(Handle, ref status);
    }
}

/// <summary>
/// A mutable pointer to network data.
/// </summary>
/// <remarks>
/// NetworkPointers allow updating references to data without changing the pointer address.
/// They are signed by the owner's key for authentication.
/// </remarks>
public sealed class NetworkPointer : NativeHandle
{
    internal NetworkPointer(IntPtr handle) : base(handle) { }

    /// <summary>
    /// Creates a new network pointer.
    /// </summary>
    /// <param name="key">The secret key to sign the pointer with.</param>
    /// <param name="counter">Version counter (increment for updates).</param>
    /// <param name="target">The target this pointer points to.</param>
    public static NetworkPointer Create(SecretKey key, ulong counter, PointerTarget target)
    {
        ArgumentNullException.ThrowIfNull(key);
        ArgumentNullException.ThrowIfNull(target);
        key.ThrowIfDisposed();
        target.ThrowIfDisposed();

        var status = RustCallStatus.Create();
        var handle = NativeMethods.NetworkPointerNew(key.CloneHandle(), counter, target.CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "NetworkPointer.Create");
        return new NetworkPointer(handle);
    }

    /// <summary>
    /// Gets the address of this pointer.
    /// </summary>
    public PointerAddress Address()
    {
        ThrowIfDisposed();
        var status = RustCallStatus.Create();
        var handle = NativeMethods.NetworkPointerAddress(CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "NetworkPointer.Address");
        return new PointerAddress(handle);
    }

    /// <summary>
    /// Gets the target this pointer points to.
    /// </summary>
    public PointerTarget Target()
    {
        ThrowIfDisposed();
        var status = RustCallStatus.Create();
        var handle = NativeMethods.NetworkPointerTarget(CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "NetworkPointer.Target");
        return new PointerTarget(handle);
    }

    /// <summary>
    /// Gets the version counter of this pointer.
    /// </summary>
    public ulong Counter
    {
        get
        {
            ThrowIfDisposed();
            var status = RustCallStatus.Create();
            return NativeMethods.NetworkPointerCounter(CloneHandle(), ref status);
        }
    }

    /// <inheritdoc/>
    protected override void FreeHandle()
    {
        var status = RustCallStatus.Create();
        NativeMethods.FreeNetworkPointer(Handle, ref status);
    }

    /// <inheritdoc/>
    protected internal override IntPtr CloneHandle()
    {
        var status = RustCallStatus.Create();
        return NativeMethods.CloneNetworkPointer(Handle, ref status);
    }
}
